File: tlog-receiver/app/src/main/java/ua/tlogreceiver/MainActivity.java
package ua.tlogreceiver;

import android.app.Activity;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Intent;
import android.database.Cursor;
import android.graphics.Color;
import android.net.Uri;
import android.os.Bundle;
import android.os.Environment;
import android.provider.MediaStore;
import android.provider.OpenableColumns;
import android.view.Gravity;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Locale;

public final class MainActivity extends Activity {
    private static final int PICK_FILE = 1001;
    private TextView status;
    private Button pick;

    @Override
    protected void onCreate(Bundle state) {
        super.onCreate(state);
        buildUi();
        if (state == null) handleIntent(getIntent());
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);
        setIntent(intent);
        handleIntent(intent);
    }

    private void buildUi() {
        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setGravity(Gravity.CENTER_HORIZONTAL);
        root.setPadding(48, 80, 48, 48);
        root.setBackgroundColor(Color.rgb(16, 20, 24));

        TextView title = new TextView(this);
        title.setText("TLOG Receiver");
        title.setTextColor(Color.WHITE);
        title.setTextSize(28);
        title.setGravity(Gravity.CENTER);
        root.addView(title, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        TextView info = new TextView(this);
        info.setText("Приймає .tlog із WhatsApp та зберігає його в Download/TLOG Receiver");
        info.setTextColor(Color.LTGRAY);
        info.setTextSize(16);
        info.setGravity(Gravity.CENTER);
        LinearLayout.LayoutParams infoLp = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        infoLp.setMargins(0, 28, 0, 40);
        root.addView(info, infoLp);

        status = new TextView(this);
        status.setText("Надішли .tlog через «Поділитися» у WhatsApp або вибери файл вручну.");
        status.setTextColor(Color.WHITE);
        status.setTextSize(16);
        status.setGravity(Gravity.CENTER);
        root.addView(status, new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        pick = new Button(this);
        pick.setText("Вибрати TLOG-файл");
        pick.setAllCaps(false);
        pick.setOnClickListener(v -> openPicker());
        LinearLayout.LayoutParams buttonLp = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, 150);
        buttonLp.setMargins(0, 40, 0, 0);
        root.addView(pick, buttonLp);

        setContentView(root);
    }

    private void openPicker() {
        Intent i = new Intent(Intent.ACTION_OPEN_DOCUMENT);
        i.addCategory(Intent.CATEGORY_OPENABLE);
        i.setType("*/*");
        startActivityForResult(i, PICK_FILE);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == PICK_FILE && resultCode == RESULT_OK && data != null && data.getData() != null) {
            save(data.getData());
        }
    }

    private void handleIntent(Intent intent) {
        if (intent == null) return;
        Uri uri = null;
        if (Intent.ACTION_SEND.equals(intent.getAction())) {
            uri = intent.getParcelableExtra(Intent.EXTRA_STREAM);
        } else if (Intent.ACTION_VIEW.equals(intent.getAction())) {
            uri = intent.getData();
        }
        if (uri != null) save(uri);
    }

    private void save(Uri source) {
        pick.setEnabled(false);
        status.setText("Зберігаю файл…");
        new Thread(() -> {
            try {
                String original = displayName(source);
                String lower = original.toLowerCase(Locale.ROOT);
                if (!(lower.endsWith(".tlog") || lower.endsWith(".bin"))) {
                    throw new IOException("Файл не має розширення .tlog або .bin: " + original);
                }
                String savedName = normalizeName(original);
                copyToDownloads(source, savedName);
                runOnUiThread(() -> {
                    status.setText("✓ Збережено: " + savedName + "\nDownload/TLOG Receiver");
                    pick.setEnabled(true);
                });
            } catch (Exception e) {
                runOnUiThread(() -> {
                    String m = e.getMessage();
                    status.setText("Помилка: " + (m == null ? "невідома помилка" : m));
                    pick.setEnabled(true);
                });
            }
        }).start();
    }

    private String displayName(Uri uri) {
        ContentResolver resolver = getContentResolver();
        try (Cursor c = resolver.query(uri, new String[]{OpenableColumns.DISPLAY_NAME}, null, null, null)) {
            if (c != null && c.moveToFirst()) {
                int i = c.getColumnIndex(OpenableColumns.DISPLAY_NAME);
                if (i >= 0) {
                    String name = c.getString(i);
                    if (name != null && !name.trim().isEmpty()) return name;
                }
            }
        } catch (Exception ignored) {}
        String segment = uri.getLastPathSegment();
        return segment == null ? "flight.tlog" : segment;
    }

    private String normalizeName(String name) {
        if (name == null || name.trim().isEmpty()) return "flight.tlog";
        String clean = name.trim();
        String lower = clean.toLowerCase(Locale.ROOT);
        if (lower.endsWith(".tlog")) return clean;
        if (lower.endsWith(".bin")) return clean.substring(0, clean.length() - 4) + ".tlog";
        return clean + ".tlog";
    }

    private void copyToDownloads(Uri source, String fileName) throws IOException {
        ContentResolver resolver = getContentResolver();
        ContentValues values = new ContentValues();
        values.put(MediaStore.MediaColumns.DISPLAY_NAME, fileName);
        values.put(MediaStore.MediaColumns.MIME_TYPE, "application/octet-stream");
        values.put(MediaStore.MediaColumns.RELATIVE_PATH, Environment.DIRECTORY_DOWNLOADS + "/TLOG Receiver");
        values.put(MediaStore.MediaColumns.IS_PENDING, 1);

        Uri dest = resolver.insert(MediaStore.Downloads.EXTERNAL_CONTENT_URI, values);
        if (dest == null) throw new IOException("Android не створив файл у Download");
        boolean ok = false;
        try (InputStream in = resolver.openInputStream(source); OutputStream out = resolver.openOutputStream(dest, "w")) {
            if (in == null || out == null) throw new IOException("Не вдалося відкрити файл");
            byte[] buf = new byte[65536];
            int n;
            while ((n = in.read(buf)) != -1) out.write(buf, 0, n);
            out.flush();
            ok = true;
        } finally {
            if (!ok) resolver.delete(dest, null, null);
        }
        ContentValues ready = new ContentValues();
        ready.put(MediaStore.MediaColumns.IS_PENDING, 0);
        resolver.update(dest, ready, null, null);
    }
}

File: android-offline/app/src/main/java/com/tloganalyzer/offline/MainActivity.kt
package com.tloganalyzer.offline

import android.annotation.SuppressLint
import android.app.Activity
import android.content.ContentValues
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.provider.MediaStore
import android.provider.OpenableColumns
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import com.chaquo.python.Python
import com.chaquo.python.android.AndroidPlatform
import org.json.JSONObject
import java.io.BufferedInputStream
import java.io.BufferedOutputStream
import java.io.File
import java.io.FileOutputStream
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

class MainActivity : Activity() {
    private lateinit var webView: WebView
    private var pendingUri: Uri? = null
    private var pageReady = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        pendingUri = extractIncomingUri(intent)
        setupWebView()
        startPythonAndOpenUi()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        pendingUri = extractIncomingUri(intent)
        if (pageReady) consumePendingShare()
    }

    @SuppressLint("SetJavaScriptEnabled")
    private fun setupWebView() {
        webView = WebView(this)
        setContentView(webView)
        webView.settings.javaScriptEnabled = true
        webView.settings.domStorageEnabled = true
        webView.settings.allowFileAccess = true
        webView.settings.allowContentAccess = true
        webView.webChromeClient = WebChromeClient()
        webView.webViewClient = object : WebViewClient() {
            override fun onPageFinished(view: WebView?, url: String?) {
                pageReady = true
                consumePendingShare()
            }
        }
    }

    private fun startPythonAndOpenUi() {
        Thread {
            try {
                if (!Python.isStarted()) Python.start(AndroidPlatform(this))
                Python.getInstance().getModule("bridge").callAttr("start_server")
                waitForServer()
                runOnUiThread { webView.loadUrl("http://127.0.0.1:8765/") }
            } catch (e: Exception) {
                showFatal("Не вдалося запустити офлайн-аналізатор: ${e.message}")
            }
        }.start()
    }

    private fun waitForServer() {
        var lastError: Exception? = null
        repeat(80) {
            try {
                val c = URL("http://127.0.0.1:8765/health").openConnection() as HttpURLConnection
                c.connectTimeout = 500
                c.readTimeout = 500
                c.requestMethod = "GET"
                if (c.responseCode == 200) {
                    c.disconnect()
                    return
                }
                c.disconnect()
            } catch (e: Exception) {
                lastError = e
            }
            Thread.sleep(250)
        }
        throw IllegalStateException("Локальний сервер не відповідає", lastError)
    }

    private fun extractIncomingUri(src: Intent?): Uri? {
        if (src == null) return null
        return when (src.action) {
            Intent.ACTION_SEND -> src.getParcelableExtra(Intent.EXTRA_STREAM)
            Intent.ACTION_VIEW -> src.data
            else -> null
        }
    }

    private fun consumePendingShare() {
        val uri = pendingUri ?: return
        pendingUri = null
        pageReady = false
        showAnalyzingState()

        Thread {
            try {
                val name = displayName(uri).ifBlank { "shared-${System.currentTimeMillis()}.tlog" }
                if (!name.lowercase().endsWith(".tlog") && !name.lowercase().endsWith(".bin")) {
                    throw IllegalArgumentException("Підтримуються лише .tlog та .bin")
                }
                val localFile = copyToCache(uri, name)
                saveToDownloads(uri, name)
                val resultJson = postForAnalysis(localFile, name)
                runOnUiThread {
                    val js = """
                        (function(){
                          try {
                            const data = $resultJson;
                            const upload = document.querySelector('.upload'); if (upload) upload.style.display='none';
                            const fileBox = document.querySelector('.file-box'); if (fileBox) fileBox.style.display='none';
                            const btn = document.querySelector('.analyze'); if (btn) btn.style.display='none';
                            if (typeof renderResults === 'function') {
                              renderResults(data);
                              window.scrollTo(0,0);
                            } else {
                              document.body.innerHTML='<pre style="color:white;padding:20px">renderResults() не знайдено</pre>';
                            }
                          } catch(e) {
                            document.body.innerHTML='<pre style="color:white;padding:20px">'+String(e)+'</pre>';
                          }
                        })();
                    """.trimIndent()
                    webView.evaluateJavascript(js, null)
                    pageReady = true
                }
            } catch (e: Exception) {
                runOnUiThread {
                    val msg = JSONObject.quote("Помилка аналізу: ${e.message}")
                    webView.evaluateJavascript("alert($msg);", null)
                    pageReady = true
                }
            }
        }.start()
    }

    private fun showAnalyzingState() {
        val js = """
            (function(){
              const p=document.querySelector('.progress-container'); if(p) p.style.display='block';
              const f=document.querySelector('.progress-fill'); if(f) f.style.width='60%';
              const h=document.querySelector('.progress-header'); if(h) h.innerHTML='<span>Офлайн-аналіз TLOG...</span><span>60%</span>';
              const u=document.querySelector('.upload'); if(u) u.style.display='none';
            })();
        """.trimIndent()
        webView.evaluateJavascript(js, null)
    }

    private fun displayName(uri: Uri): String {
        contentResolver.query(uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null)?.use { c ->
            if (c.moveToFirst()) {
                val i = c.getColumnIndex(OpenableColumns.DISPLAY_NAME)
                if (i >= 0) return c.getString(i) ?: ""
            }
        }
        return uri.lastPathSegment?.substringAfterLast('/') ?: ""
    }

    private fun copyToCache(uri: Uri, name: String): File {
        val dir = File(cacheDir, "shared").apply { mkdirs() }
        val out = File(dir, name.replace(Regex("[^A-Za-z0-9._ -]"), "_"))
        contentResolver.openInputStream(uri).use { input ->
            requireNotNull(input) { "Не вдалося прочитати файл" }
            FileOutputStream(out).use { output -> input.copyTo(output, 1024 * 128) }
        }
        return out
    }

    private fun saveToDownloads(uri: Uri, name: String) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) return
        val values = ContentValues().apply {
            put(MediaStore.Downloads.DISPLAY_NAME, name)
            put(MediaStore.Downloads.MIME_TYPE, "application/octet-stream")
            put(MediaStore.Downloads.RELATIVE_PATH, Environment.DIRECTORY_DOWNLOADS + "/TLOG Receiver")
            put(MediaStore.Downloads.IS_PENDING, 1)
        }
        val target = contentResolver.insert(MediaStore.Downloads.EXTERNAL_CONTENT_URI, values) ?: return
        try {
            contentResolver.openInputStream(uri).use { input ->
                contentResolver.openOutputStream(target).use { output ->
                    if (input != null && output != null) input.copyTo(output, 1024 * 128)
                }
            }
            values.clear()
            values.put(MediaStore.Downloads.IS_PENDING, 0)
            contentResolver.update(target, values, null, null)
        } catch (e: Exception) {
            contentResolver.delete(target, null, null)
        }
    }

    private fun postForAnalysis(file: File, originalName: String): String {
        val boundary = "----TLOG-${UUID.randomUUID()}"
        val c = URL("http://127.0.0.1:8765/analyze").openConnection() as HttpURLConnection
        c.requestMethod = "POST"
        c.doOutput = true
        c.connectTimeout = 10_000
        c.readTimeout = 240_000
        c.setChunkedStreamingMode(1024 * 128)
        c.setRequestProperty("Content-Type", "multipart/form-data; boundary=$boundary")

        BufferedOutputStream(c.outputStream).use { out ->
            out.write("--$boundary\r\n".toByteArray())
            out.write("Content-Disposition: form-data; name=\"file\"; filename=\"$originalName\"\r\n".toByteArray())
            out.write("Content-Type: application/octet-stream\r\n\r\n".toByteArray())
            BufferedInputStream(file.inputStream()).use { input -> input.copyTo(out, 1024 * 128) }
            out.write("\r\n--$boundary--\r\n".toByteArray())
            out.flush()
        }

        val code = c.responseCode
        val stream = if (code in 200..299) c.inputStream else c.errorStream
        val text = stream.bufferedReader().use { it.readText() }
        c.disconnect()
        if (code !in 200..299) throw IllegalStateException("HTTP $code: $text")
        return text
    }

    private fun showFatal(message: String) {
        runOnUiThread {
            webView.loadDataWithBaseURL(
                null,
                "<html><body style='background:#0a0d11;color:white;font-family:sans-serif;padding:24px'><h2>TLOG Analyzer Offline</h2><p>${message}</p></body></html>",
                "text/html",
                "UTF-8",
                null
            )
        }
    }
}
